File: leetcode437/leetcode437.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
               def dfs(node, presum):
                     if not node: return 0
                     presum += node.val

                     path_cnt = presum_counts.get(presum - targetSum, 0)
                     presum_counts[presum] = presum_counts.get(presum, 0) + 1
                     path_cnt += dfs(node.left, presum) + dfs(node.right, presum)
                     presum_counts[presum] -= 1     
                     
                     return path_cnt
               
               presum_counts = {0 : 1}
               return dfs(root, 0)

File: leetcode437/leetcode437.cpp
#include<bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    unordered_map<long long, int> preSumCount;

public:
    int dfs(TreeNode* node, int targetSum, long long preSum){
        if (!node) return 0;
        preSum += node->val;
        int pathCnt = preSumCount[preSum - targetSum];

        preSumCount[preSum] += 1;
        pathCnt += dfs(node->left, targetSum, preSum) + dfs(node->right, targetSum, preSum);

        preSumCount[preSum] -= 1;
        return pathCnt;
    }
    int pathSum(TreeNode* root, int targetSum) {
        preSumCount[0] = 1;
        return dfs(root, targetSum, 0);
    }
};